### ISSUE TITLE
Loss nan

### ISSUE THREAD

**Issue by cindybrain** [NervanaSystems/distiller#487]
When I run the command:
```python3 compress_classifier.py -a=resnet18 ../../../data.imagenet -j 22 --epoch=180 --lr=0.001 --compress=../../../yaml/resnet18-baseline.yaml --gpu=0 ```
resnet18-baseline.yaml:
```
lr_schedulers:
  training_lr:
    class: StepLR
    step_size: 45
    gamma: 0.10

policies:
    - lr_scheduler:
        instance_name: training_lr
      starting_epoch: 45
      ending_epoch: 200
      frequency: 1
```
At first, the training run all right. But at epoch 19, the loss changed to nan, while the correctness began to deline. I don't know whether my data sets have something wrong or  lr or something else.

fragment of the log:
```
2020-03-04 19:31:07,350 - Epoch: [19][ 2290/ 4505]    Overall Loss 3.239098    Objective Loss 3.239098    Top1 32.776508    Top5 57.088735    LR 0.001000    Time 0.607148    
2020-03-04 19:31:19,060 - Epoch: [19][ 2300/ 4505]    Overall Loss 3.239080    Objective Loss 3.239080    Top1 32.776664    Top5 57.089164    LR 0.001000    Time 0.609587    
2020-03-04 19:31:21,718 - Epoch: [19][ 2310/ 4505]    Overall Loss 3.238959    Objective Loss 3.238959    Top1 32.782738    Top5 57.090605    LR 0.001000    Time 0.608092    
2020-03-04 19:31:34,043 - Epoch: [19][ 2320/ 4505]    Overall Loss 3.238968    Objective Loss 3.238968    Top1 32.783371    Top5 57.090349    LR 0.001000    Time 0.610778    
2020-03-04 19:31:36,740 - Epoch: [19][ 2330/ 4505]    Overall Loss 3.239078    Objective Loss 3.239078    Top1 32.788358    Top5 57.088754    LR 0.001000    Time 0.609309    
2020-03-04 19:31:45,462 - Epoch: [19][ 2340/ 4505]    Overall Loss 3.239078    Objective Loss 3.239078    Top1 32.790465    Top5 57.085837    LR 0.001000    Time 0.610432    
2020-03-04 19:31:48,100 - Epoch: [19][ 2350/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.725066    Top5 56.964262    LR 0.001000    Time 0.608939    
2020-03-04 19:31:59,121 - Epoch: [19][ 2360/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.586566    Top5 56.725040    LR 0.001000    Time 0.611020    
2020-03-04 19:32:01,889 - Epoch: [19][ 2370/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.449400    Top5 56.487507    LR 0.001000    Time 0.609588    
2020-03-04 19:32:07,246 - Epoch: [19][ 2380/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.313715    Top5 56.253283    LR 0.001000    Time 0.609252    
2020-03-04 19:32:15,716 - Epoch: [19][ 2390/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.178511    Top5 56.018894    LR 0.001000    Time 0.610237    
2020-03-04 19:32:18,554 - Epoch: [19][ 2400/ 4505]    Overall Loss nan    Objective Loss nan    Top1 32.044759    Top5 55.787760    LR 0.001000    Time 0.608862    
```
I want to know whether anyone else met the same problem. Thank you:)

**Comment by nzmora** [comment]
Hi @cindybrain ,

I have not seen this issue before.
A couple of things I can point out:
1. The problem starts at epoch 19, so it should not have anything to do with `resnet18-baseline.yaml`, because that is only scheduled at epoch 45.
2. If the loss is a nan, then one of the loss components is problematic.  So try to break when the loss is nan (use `torch.isnan`).  Which layer is introducing the problem?  Is it a conv or fc layer?  Are the weights problematic?  If so, is the problem with the gradients or the optimizer update?

I suspect the gradients (vanishing or exploding), in which case try changing the initialization of the weights.  For conv layers, pytorch opts for `init.kaiming_uniform_` by default:

https://github.com/pytorch/pytorch/blob/9a5e9d8cecb1c68eebd69618729a7ec5671b2a06/torch/nn/modules/conv.py#L56

Try `nn.init.kaiming_normal_`

Cheers
Neta

**Comment by zeyu-hello** [comment]
I encoutered the same probelm while use this command line:
```
python compress_detector.py --data-path ./public_data/coco --compress maskrcnn.scheduler_agp_structure.yaml
```

My config file:
```
version: 1

pruners:

  # fc_pruner:
  #   class: AutomatedGradualPruner
  #   initial_sparsity : 0.01
  #   final_sparsity: 0.85
  #   weights: [
  #     roi_heads.box_head.fc6.weight,
  #     roi_heads.box_head.fc7.weight
  #   ]

  agp_pruner_75:
    class: L1RankedStructureParameterPruner_AGP
    initial_sparsity : 0.01
    final_sparsity: 0.75
    group_type: Filters
    weights: [
      backbone.body.layer1.0.conv1.weight,
      backbone.body.layer1.0.conv2.weight,
      backbone.body.layer1.0.conv3.weight,
      backbone.body.layer1.0.downsample.0.weight,
      backbone.body.layer1.1.conv1.weight,
      backbone.body.layer1.1.conv2.weight,
      backbone.body.layer1.1.conv3.weight,
      backbone.body.layer1.2.conv1.weight,
      backbone.body.layer1.2.conv2.weight,
      backbone.body.layer1.2.conv3.weight,
      backbone.body.layer2.0.conv1.weight,
      backbone.body.layer2.0.conv2.weight,
      backbone.body.layer2.0.conv3.weight,]

  agp_pruner_85:
    class: L1RankedStructureParameterPruner_AGP
    initial_sparsity : 0.01
    final_sparsity: 0.85
    group_type: Filters
    weights: [
      backbone.body.layer2.0.downsample.0.weight,
      backbone.body.layer2.1.conv1.weight,
      backbone.body.layer2.1.conv2.weight,
      backbone.body.layer2.1.conv3.weight,
      backbone.body.layer2.2.conv1.weight,
      backbone.body.layer2.2.conv2.weight,
      backbone.body.layer2.2.conv3.weight,
      backbone.body.layer2.3.conv1.weight,
      backbone.body.layer2.3.conv2.weight,
      backbone.body.layer2.3.conv3.weight,
      backbone.body.layer3.0.conv1.weight,
      backbone.body.layer3.0.conv2.weight,
      backbone.body.layer3.0.conv3.weight,
      backbone.body.layer3.0.downsample.0.weight,
      backbone.body.layer3.1.conv1.weight,
      backbone.body.layer3.1.conv2.weight,
      backbone.body.layer3.1.conv3.weight,
      backbone.body.layer3.2.conv1.weight,
      backbone.body.layer3.2.conv2.weight,
      backbone.body.layer3.2.conv3.weight,
      backbone.body.layer3.3.conv1.weight,
      backbone.body.layer3.3.conv2.weight,
      backbone.body.layer3.3.conv3.weight,
      backbone.body.layer3.4.conv1.weight,
      backbone.body.layer3.4.conv2.weight,
      backbone.body.layer3.4.conv3.weight,
      backbone.body.layer3.5.conv1.weight,
      backbone.body.layer3.5.conv2.weight,
      backbone.body.layer3.5.conv3.weight,
      backbone.body.layer4.2.conv3.weight,
      # backbone.fpn.inner_blocks.0.weight,
      # backbone.fpn.inner_blocks.1.weight,
      # backbone.fpn.inner_blocks.2.weight,
      # backbone.fpn.inner_blocks.3.weight,
      # backbone.fpn.layer_blocks.0.weight,
      # backbone.fpn.layer_blocks.1.weight,
      # backbone.fpn.layer_blocks.2.weight,
      # backbone.fpn.layer_blocks.3.weight,
      # rpn.head.conv.weight,
      # roi_heads.mask_head.mask_fcn1.weight,
      # roi_heads.mask_head.mask_fcn2.weight,
      # roi_heads.mask_head.mask_fcn3.weight,
      # roi_heads.mask_head.mask_fcn4.weight,
      # roi_heads.mask_predictor.conv5_mask.weight,
    ]

  agp_pruner_90:
    class: L1RankedStructureParameterPruner_AGP
    initial_sparsity : 0.01
    final_sparsity: 0.90
    group_type: Filters
    weights: [
    backbone.body.layer4.0.conv1.weight,
    backbone.body.layer4.0.conv2.weight,
    backbone.body.layer4.0.conv3.weight,
    backbone.body.layer4.0.downsample.0.weight,
    backbone.body.layer4.1.conv1.weight,
    backbone.body.layer4.1.conv2.weight,
    backbone.body.layer4.1.conv3.weight,
    backbone.body.layer4.2.conv1.weight,
    backbone.body.layer4.2.conv2.weight,
    ]


policies:
  - pruner:
     instance_name : agp_pruner_75
    starting_epoch: 0
    ending_epoch: 45
    frequency: 1

  - pruner:
     instance_name : agp_pruner_85
    starting_epoch: 0
    ending_epoch: 45
    frequency: 1

  # - pruner:
  #     instance_name : fc_pruner
  #   starting_epoch: 0
  #   ending_epoch: 45
  #   frequency: 3

  - pruner:
      instance_name : agp_pruner_90
    starting_epoch: 0
    ending_epoch: 45
    frequency: 1
```
